### PR TITLE
CVAT push ToolTracking active learning model to Nuclio function container

### PR DESCRIPTION
**Pull request by tangy5**
Signed-off-by: tangy5 <yucheng.tang@vanderbilt.edu>

Hi @SachidanandAlle , 

I've made the script for pushing fine-tuned tooltracking model to CVAT nuclio function container, and to replace the model for inference.

The script works as follow:

1. Check if the function container is up and running.
2. Check if the latest fine-tuned model is published/saved in MONAI Label Endoscopy workspace in /model folder.
3. Push and replace the pre-trained model in CVAT Nuclio function container.
4. The user can inference with the latest published model for labeling. 

Readme file is changed a bit according to the operation. 

Let me know if this is the desired thing for now. Users need to run the commands in .sh to push the latest model.
We may need the automated function for pushing the model to CVAT Nuclio Container once MONAI Label server training is done in future.

### CODE REVIEW

File: sample-apps/endoscopy/pushActiveLearningModel.sh [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+MODEL_CONTAINER="/opt/conda/monailabel/sample-apps/endoscopy/model/pretrained_$FUNC_NAME.pt" # default model path at function container
+
+# Check if latest active learning model checkpoint is done and saved.
+if [ ! -f $CKP_FILE ]; then
+    echo "Active learning model checkpoint not published, exiting..."
+else
+    # Replace prior pretrained model with lastest model as current pre-trained model
+    $(docker cp "$CKP_FILE" "$FUNC_CONTAINER:$MODEL_CONTAINER")
+    echo "Published latest mode: $CKP_FILE into $FUNC_NAME nuclio function container..."
+fi
```
Comment (Nic-Ma) [inline sample-apps/endoscopy/pushActiveLearningModel.sh:42]: missing a blank line.

Thanks.

File: sample-apps/endoscopy/README.md [hunk 2]
```diff
@@ -109,6 +106,11 @@ monailabel start_server \
   --conf epistemic_top_k 3 \
   --conf auto_finetune_models tooltracking \
   --conf auto_finetune_check_interval 30
+
+# Active Learning for Inference
+# After fine-tuned model is saved, push the model to nuclio function container
+/workspace-apps/endoscopy/pushActiveLearningModel.sh
```
Comment (Nic-Ma) [inline sample-apps/endoscopy/README.md:112]: I think should add the CLI command? For example:
`bash /workspace-apps/endoscopy/pushActiveLearningModel.sh`

Thanks.

**Review by Nic-Ma — COMMENTED** [review]
Hi @SachidanandAlle ,

Could you please help also take a look at this PR?
I see there is no unit tests to cover it, is it OK?

Thanks.

**Comment by SachidanandAlle** [comment]
Thanks.. I will look into this..

**Comment by SachidanandAlle** [comment]
Let's think of a pull paradigm instead of push..

Introduce monailabel endpoints get details about latest model available.. checksum etc.. and consumers can fetch it through APIs

**Comment by SachidanandAlle** [comment]
For this release let it be this way.. we can polish the script and arguments

**Comment by tangy5** [comment]
Thanks! Will work on this to update.

File: sample-apps/endoscopy/README.md (hunk 2 above)
Comment (tangy5) [reply sample-apps/endoscopy/README.md]: Thanks

File: sample-apps/endoscopy/README.md [hunk 3]
```diff
@@ -131,6 +128,11 @@ monailabel start_server \
   --conf epistemic_top_k 3 \
   --conf auto_finetune_models tooltracking \
   --conf auto_finetune_check_interval 30
+
+# Active Learning for Inference
+# After fine-tuned model is saved, push the model to nuclio function container
+bash /workspace-apps/endoscopy/bin/pushActiveLearningModel.sh
```
Comment (SachidanandAlle) [inline sample-apps/endoscopy/README.md:134]: workspace/endoscopy/bin/xyz.sh

**Comment by SachidanandAlle** [comment]
.gitignore.swp needs to be removed + script.sh file is missing

File: sample-apps/endoscopy/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (Nic-Ma) [reply sample-apps/endoscopy/README.md]: Hi @SachidanandAlle ,

I think if run with `bash XXX`, don't need +x permission?

Thanks.

File: sample-apps/endoscopy/pushActiveLearningModel.sh [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+# Sample commands to publish active learning trained latest model to nuclio functions
+# ./pushActiveLearningModel.sh
+
+SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" >/dev/null 2>&1 && pwd)"
+
+FUNCTIONS_DIR=${2:-$SCRIPT_DIR}
+
+# only support tooltracking for now, deid,
+# to add loop checks for all containers/functions and published models, then push to containers.
+FUNC_NAME="tooltracking"
```
[the remark this answers is not in the document]
Comment (tangy5) [reply sample-apps/endoscopy/pushActiveLearningModel.sh]: Done

[the remark this answers is not in the document]
Comment (tangy5) [reply sample-apps/endoscopy/pushActiveLearningModel.sh]: Done

File: sample-apps/endoscopy/README.md (hunk 3 above)
Comment (tangy5) [reply sample-apps/endoscopy/README.md]: Thanks, fixed.

**Comment by tangy5** [comment]
> .gitignore.swp needs to be removed + script.sh file is missing

Changed ./gitignore, commented #bin. Pushed update_cvat_model.sh

Thanks, others are adjusted according to suggestions. The script is tested with latest MONAI rc2, Endoscopy app.

File: .gitignore [hunk 5]
```diff
@@ -161,7 +161,7 @@ docs/source/apidocs
 *.nii.gz
 .DS_Store
 *.DS_Store
-bin
+#bin
```
Comment (SachidanandAlle) [inline .gitignore:164]: lets restore this.. any files u add you should use git add -f option

File: sample-apps/endoscopy/bin/update_cvat_model.sh [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+# Check if latest active learning model checkpoint is done and saved, error if blank.
+if [ -z "$MODEL_PATH" ] || [ ! -f "$MODEL_PATH" ]; then
+    echo "Active learning model checkpoint not provided or published, exiting..."
+else
+    # Replace prior pretrained model with lastest model as current pre-trained model
+    $(docker cp "$MODEL_PATH" "$FUNC_CONTAINER:$MODEL_CONTAINER")
+    echo "Published latest mode: $MODEL_PATH into $FUNC_NAME nuclio function container."
+fi
+kkjkjj
```
Comment (SachidanandAlle) [inline sample-apps/endoscopy/bin/update_cvat_model.sh:44]: kkjkkjj?

File: sample-apps/endoscopy/README.md [hunk 7]
```diff
@@ -131,6 +128,12 @@ monailabel start_server \
   --conf epistemic_top_k 3 \
   --conf auto_finetune_models tooltracking \
   --conf auto_finetune_check_interval 30
+
+# Active Learning for Inference
```
Comment (SachidanandAlle) [inline sample-apps/endoscopy/README.md:132]: Comment is something like this... (not really active learning for inference)
# Update/publish latest model back to CVAT

File: sample-apps/endoscopy/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (SachidanandAlle) [reply sample-apps/endoscopy/README.md]: sometimes we don't assume.. and people normally try ./ instead of bash or sh

File: sample-apps/endoscopy/bin/update_cvat_model.sh (hunk 6 above)
Comment (tangy5) [reply sample-apps/endoscopy/bin/update_cvat_model.sh:44]: Fixed.

File: sample-apps/endoscopy/README.md (hunk 7 above)
Comment (tangy5) [reply sample-apps/endoscopy/README.md]: Done

File: .gitignore (hunk 5 above)
Comment (tangy5) [reply .gitignore]: Ok, done.

**Review by SachidanandAlle — APPROVED** [review]

**Comment by tangy5** [comment]
Thanks for all the help, Sachi, very appreciative of this.
